dot/core/consensus.go

```diff
@@ -0,0 +1,177 @@
+		StorageState:     s.storageState,
+		TransactionQueue: s.transactionQueue,
+		AuthData:         authData,
+		Done:             epochDone,
+		Kill:             babeKill,
+		StartSlot:        bestSlot + 1,
+		SyncLock:         s.syncLock,
+	}
+
+	// create new BABE session
+	bs, err := babe.NewSession(bsConfig)
+	if err != nil {
+		log.Error("[core] failed to initialize BABE session", "error", err)
+		return nil, err
+	}
+
+	log.Debug(
+		"[core] BABE session initialized",
+		"epoch", s.epochNumber,
+	)
+
+	return bs, nil
+}
+
+// handleConsensusDigest checks if the provided header is the block header for
+// the first block of the current epoch, finds and decodes the consensus digest
+// item from the block digest, and then sets the epoch data for the next epoch
+func (s *Service) handleConsensusDigest(header *types.Header) (err error) {
+
+	// check if block header digest items exist
+	if header.Digest == nil || len(header.Digest) == 0 {
+		return fmt.Errorf("header digest not set")
+	}
+
+	// declare digest item
+	var item types.DigestItem
+
+	// decode each digest item and check its type
+	for _, digest := range header.Digest {
+		item, err = types.DecodeDigestItem(digest)
+		if err != nil {
+			return err
+		}
+
+		// check if digest item is consensus digest type
+		if item.Type() == types.ConsensusDigestType {
+
+			// cast decoded consensus digest item to conensus digest type
+			consensusDigest := item.(*types.ConsensusDigest)
+
+			// initialize epoch data interface for next epoch
+			nextEpochData := new(babe.NextEpochDescriptor)
+
+			// decode consensus digest data for next epoch
+			err = nextEpochData.Decode(consensusDigest.Data)
+			if err != nil {
+				return err
+			}
+
+			// set epcoh data for next epoch
+			err = s.bs.SetEpochData(nextEpochData)
+			if err != nil {
+				return err
+			}
+
+			// set first block in current epoch
+			s.firstBlock = header.Number
+		}
+	}
+
+	return nil
+}
+
+// verifyFirstBlock verifies first block has already been set for current epoch,
+// therefore, consensus digest was found and epoch data for next epoch was set
+func (s *Service) verifyFirstBlock() error {
+
+	// get first block in current epoch
+	block, err := s.blockState.GetBlockByNumber(s.firstBlock)
+	if err != nil {
+		return fmt.Errorf("failed to get first block: %s", err)
+	}
+
+	// verify first block is current epoch
+	err = s.verifyCurrentEpoch(block.Header)
+	if err != nil {
+		return fmt.Errorf("first block not set for current epoch: %s", err)
+	}
+
+	return nil
+}
+
+// verifyCurrentEpoch verifies block header is from current epoch
+func (s *Service) verifyCurrentEpoch(header *types.Header) error {
+
+	// get slot number to determine epoch number
+	slot, err := s.blockState.GetSlotForBlock(header.Hash())
+	if err != nil {
+		return fmt.Errorf("failed to get slot for block: %s", err)
+	}
+
+	// epoch number = (slot - genesis slot) / epoch length
+	epoch := (slot - 0) / 6 // TODO: set epoch length in core config
```

Comment:
add a helper method to babe so that the epoch length can be accessed with `s.bs.EpochLength()` or something similar?

Reply:
also `slot - 0` ? can you remove the ` -0`

Reply:
that would be nice, was not sure how to do this and did not want to extend this pull request beyond the core service. I can create an issue for this.

Reply:
I kept `slot - 0` because `slot - genesis slot`, which I think we should keep and replace with a constant or param but was not sure if we were storing the genesis slot somewhere. I think it makes sense to keep `slot - 0` given we are assuming genesis slot is `0`.

Reply:
the first slot should be slot 1 I believe

Reply:
do you mean genesis slot? should it be a constant? should it be configurable?

Reply:
it should always be 1 so this line should be `(slot - 1) / 6`

then slots 1-6 are in epoch 0, 7-12 in epoch 1. etc